<a href="https://colab.research.google.com/github/michaelryanka/ASG-PYTN-H8-Batch-062/blob/main/PYTN_Assgn_2_Batch_062_Muhammad_Michael_Ryanka.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Title: Hacktiv8 PTP Introduction to Data Science Projects 2: Statistical Treatment for Datasets Starter Notebook\
Last Updated: September 20, 2020\
Author: Raka Ardhi

## NYC Property Sales Introduction

The aim of this projects is to introduce you to practical statistic with Python as concrete and as consistent as possible. Using what you’ve learned; download the NYC Property Sales Dataset from Kaggle. This dataset is a record of every building or building unit (apartment, etc.) sold in the New York City property market over a 12-month period.

This dataset contains the location, address, type, sale price, and sale date of building units sold. A reference on the trickier fields:

* `BOROUGH`: A digit code for the borough the property is located in; in order these are Manhattan (1), Bronx (2), Brooklyn (3), Queens (4), and Staten Island (5).
* `BLOCK`; `LOT`: The combination of borough, block, and lot forms a unique key for property in New York City. Commonly called a BBL.
* `BUILDING CLASS AT PRESENT` and `BUILDING CLASS AT TIME OF SALE`: The type of building at various points in time.

Note that because this is a financial transaction dataset, there are some points that need to be kept in mind:

* Many sales occur with a nonsensically small dollar amount: $0 most commonly. These sales are actually transfers of deeds between parties: for example, parents transferring ownership to their home to a child after moving out for retirement.
* This dataset uses the financial definition of a building/building unit, for tax purposes. In case a single entity owns the building in question, a sale covers the value of the entire building. In case a building is owned piecemeal by its residents (a condominium), a sale refers to a single apartment (or group of apartments) owned by some individual.

Formulate a question and derive a statistical hypothesis test to answer the question. You have to demonstrate that you’re able to make decisions using data in a scientific manner. Examples of questions can be:

* Is there a difference in unit sold between property built in 1900-2000 and 2001 so on?
* Is there a difference in unit sold based on building category?
* What can you discover about New York City real estate by looking at a year's worth of raw transaction records? Can you spot trends in the market?

Please make sure that you have completed the lesson for this course, namely Python and Practical Statistics which is part of this Program.

**Note:** You can take a look at Project Rubric below:

| Code Review |  |
| :--- | :--- |
| CRITERIA | SPECIFICATIONS |
| Mean | Student implement mean to specifics column/data using pandas, numpy, or scipy|
| Median | Student implement median to specifics column/data using pandas, numpy, or scipy|
| Modus | Student implement modus to specifics column/data using pandas, numpy, or scipy|
| Central Tendencies | Implementing Central Tendencies through dataset |
| Box Plot | Implementing Box Plot to visualize spesific data |
| Z-Score | Implementing Z-score concept to specific data |
| Probability Distribution | Student analyzing distribution of data and gain insight from the distribution |
| Intervals | Implementing Confidence or Prediction Intervals |
| Hypotesis Testing | Made 1 Hypotesis and get conclusion from data |
| Preprocessing | Student preprocess dataset before applying the statistical treatment. |
| Does the code run without errors? | The code runs without errors. All code is functional and formatted properly. |

| Readability |  |
| :--- | :--- |
| CRITERIA | SPECIFICATIONS |
| Well Documented | All cell in notebook are well documented with markdown above each cell explaining the code|

| Analysis |  |
| :--- | :--- |
| CRITERIA | SPECIFICATIONS |
|Overall Analysis| Gain an insight/conclusion of overall plots that answer the hypotesis |

**Focus on "Graded-Function" sections.**

------------

## Data Preparation

Load the library you need.

Get your NYC property data from [here](https://www.kaggle.com/new-york-city/nyc-property-sales) and load the dataframe to your notebook.

In [ ]:
# Get your import statement here
import math
import statistics
import numpy as np
import scipy.stats
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt


%matplotlib inline
import matplotlib.pyplot as plt
from IPython.display import Math, Latex
#from IPyhton.display import Image

import seaborn as sns

In [ ]:
# Load your dataset here
df= pd.read_csv('nyc-rolling-sales.csv')
print ('Data read into a pandas dataframe!')

Data read into a pandas dataframe!


Let's view the top 5 rows of the dataset using the `head()` function.

In [ ]:
# Write your syntax here
df.head()

Unnamed: 0  BOROUGH  ... SALE PRICE            SALE DATE
0           4        1  ...    6625000  2017-07-19 00:00:00
1           5        1  ...        -    2016-12-14 00:00:00
2           6        1  ...        -    2016-12-09 00:00:00
3           7        1  ...    3936272  2016-09-23 00:00:00
4           8        1  ...    8000000  2016-11-17 00:00:00

[5 rows x 22 columns]

We can also veiw the bottom 5 rows of the dataset using the `tail()` function.

In [ ]:
# Write your syntax here
df.tail()

Unnamed: 0  BOROUGH  ... SALE PRICE            SALE DATE
84543        8409        5  ...     450000  2016-11-28 00:00:00
84544        8410        5  ...     550000  2017-04-21 00:00:00
84545        8411        5  ...     460000  2017-07-05 00:00:00
84546        8412        5  ...   11693337  2016-12-21 00:00:00
84547        8413        5  ...      69300  2016-10-27 00:00:00

[5 rows x 22 columns]

BOROUGH: A digit code for the borough the property is located in; in order these are Manhattan (1), Bronx (2), Brooklyn (3), Queens (4), and Staten Island (5).

To view the dimensions of the dataframe, we use the `.shape` parameter. Expected result: (84548, 22)

In [ ]:
# Write your syntax here
df.shape

(84548, 22)

According to this official page, Ease-ment is "is a right, such as a right of way, which allows an entity to make limited use of another’s real property. For example: MTA railroad tracks that run across a portion of another property". Also, the Unnamed column is not mentioned and was likely used for iterating through records. So, those two columns are removed for now.

In [ ]:
# Drop 'Unnamed: 0' and 'EASE-MENT' features using .drop function

df.drop(columns={"Unnamed: 0","EASE-MENT"}, inplace=True)
df

BOROUGH   NEIGHBORHOOD  ... SALE PRICE            SALE DATE
0            1  ALPHABET CITY  ...    6625000  2017-07-19 00:00:00
1            1  ALPHABET CITY  ...        -    2016-12-14 00:00:00
2            1  ALPHABET CITY  ...        -    2016-12-09 00:00:00
3            1  ALPHABET CITY  ...    3936272  2016-09-23 00:00:00
4            1  ALPHABET CITY  ...    8000000  2016-11-17 00:00:00
...        ...            ...  ...        ...                  ...
84543        5        WOODROW  ...     450000  2016-11-28 00:00:00
84544        5        WOODROW  ...     550000  2017-04-21 00:00:00
84545        5        WOODROW  ...     460000  2017-07-05 00:00:00
84546        5        WOODROW  ...   11693337  2016-12-21 00:00:00
84547        5        WOODROW  ...      69300  2016-10-27 00:00:00

[84548 rows x 20 columns]

Let's view Dtype of each features in dataframe using `.info()` function.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84548 entries, 0 to 84547
Data columns (total 20 columns):
 #   Column                          Non-Null Count  Dtype 
---  ------                          --------------  ----- 
 0   BOROUGH                         84548 non-null  int64 
 1   NEIGHBORHOOD                    84548 non-null  object
 2   BUILDING CLASS CATEGORY         84548 non-null  object
 3   TAX CLASS AT PRESENT            84548 non-null  object
 4   BLOCK                           84548 non-null  int64 
 5   LOT                             84548 non-null  int64 
 6   BUILDING CLASS AT PRESENT       84548 non-null  object
 7   ADDRESS                         84548 non-null  object
 8   APARTMENT NUMBER                84548 non-null  object
 9   ZIP CODE                        84548 non-null  int64 
 10  RESIDENTIAL UNITS               84548 non-null  int64 
 11  COMMERCIAL UNITS                84548 non-null  int64 
 12  TOTAL UNITS                     84548 non-null

It looks like empty records are not being treated as NA. We convert columns to their appropriate data types to obtain NAs.

In [ ]:
#First, let's check which columns should be categorical
print('Column name')
for col in df.columns:
    if df[col].dtype=='object':
        print(col, df[col].nunique())

Column name
NEIGHBORHOOD 254
BUILDING CLASS CATEGORY 47
TAX CLASS AT PRESENT 11
BUILDING CLASS AT PRESENT 167
ADDRESS 67563
APARTMENT NUMBER 3989
LAND SQUARE FEET 6062
GROSS SQUARE FEET 5691
BUILDING CLASS AT TIME OF SALE 166
SALE PRICE 10008
SALE DATE 364


In [ ]:
# LAND SQUARE FEET,GROSS SQUARE FEET, SALE PRICE, BOROUGH should be numeric. 
# SALE DATE datetime format.
# categorical: NEIGHBORHOOD, BUILDING CLASS CATEGORY, TAX CLASS AT PRESENT, BUILDING CLASS AT PRESENT,
# BUILDING CLASS AT TIME OF SALE, TAX CLASS AT TIME OF SALE,BOROUGH 

numer = ['LAND SQUARE FEET','GROSS SQUARE FEET', 'SALE PRICE', 'BOROUGH']
for col in numer: # coerce for missing values
    df[col] = pd.to_numeric(df[col], errors='coerce')

categ = ['NEIGHBORHOOD', 'BUILDING CLASS CATEGORY', 'TAX CLASS AT PRESENT', 'BUILDING CLASS AT PRESENT', 'BUILDING CLASS AT TIME OF SALE', 'TAX CLASS AT TIME OF SALE']
for col in categ:
    df[col] = df[col].astype('category')

df['SALE DATE'] = pd.to_datetime(df['SALE DATE'], errors='coerce')

Our dataset is ready for checking missing values.

In [ ]:
missing = df.isnull().sum()/len(df)*100

print(pd.DataFrame([missing[missing>0],pd.Series(df.isnull().sum()[df.isnull().sum()>1000])], index=['percent missing','how many missing']))

                  LAND SQUARE FEET  GROSS SQUARE FEET   SALE PRICE
percent missing          31.049818          32.658372     17.22217
how many missing      26252.000000       27612.000000  14561.00000


Around 30% of GROSS SF and LAND SF are missing. Furthermore, around 17% of SALE PRICE is also missing.

We can fill in the missing value from one column to another, which will help us reduce missing values. Expected values:

(6, 20)

(1366, 20)

In [ ]:
print(df[(df['LAND SQUARE FEET'].isnull()) & (df['GROSS SQUARE FEET'].notnull())].shape)
print(df[(df['LAND SQUARE FEET'].notnull()) & (df['GROSS SQUARE FEET'].isnull())].shape)

(6, 20)
(1366, 20)


There are 1372 rows that can be filled in with their approximate values.

In [ ]:
df['LAND SQUARE FEET'] = df['LAND SQUARE FEET'].mask((df['LAND SQUARE FEET'].isnull()) & (df['GROSS SQUARE FEET'].notnull()), df['GROSS SQUARE FEET'])
df['GROSS SQUARE FEET'] = df['GROSS SQUARE FEET'].mask((df['LAND SQUARE FEET'].notnull()) & (df['GROSS SQUARE FEET'].isnull()), df['LAND SQUARE FEET'])

In [ ]:
#  Check for duplicates before

print(sum(df.duplicated()))

df[df.duplicated(keep=False)].sort_values(['NEIGHBORHOOD', 'ADDRESS']).head(10)

# df.duplicated() automatically excludes duplicates, to keep duplicates in df we use keep=False

# in df.duplicated(df.columns) we can specify column names to look for duplicates only in those mentioned columns.

765


BOROUGH   NEIGHBORHOOD  ... SALE PRICE  SALE DATE
76286        5       ANNADALE  ...        NaN 2017-06-27
76287        5       ANNADALE  ...        NaN 2017-06-27
76322        5       ANNADALE  ...        NaN 2017-05-11
76323        5       ANNADALE  ...        NaN 2017-05-11
76383        5  ARDEN HEIGHTS  ...   320000.0 2017-06-06
76384        5  ARDEN HEIGHTS  ...   320000.0 2017-06-06
76643        5       ARROCHAR  ...        NaN 2017-03-21
76644        5       ARROCHAR  ...        NaN 2017-03-21
50126        4        ASTORIA  ...        NaN 2017-01-12
50127        4        ASTORIA  ...        NaN 2017-01-12

[10 rows x 20 columns]

The dataframe has 765 duplicated rows (exluding the original rows).

In [ ]:

df.drop_duplicates(inplace=True)

print(sum(df.duplicated()))

0


In [ ]:
df.isnull().sum()

BOROUGH                               0
NEIGHBORHOOD                          0
BUILDING CLASS CATEGORY               0
TAX CLASS AT PRESENT                  0
BLOCK                                 0
LOT                                   0
BUILDING CLASS AT PRESENT             0
ADDRESS                               0
APARTMENT NUMBER                      0
ZIP CODE                              0
RESIDENTIAL UNITS                     0
COMMERCIAL UNITS                      0
TOTAL UNITS                           0
LAND SQUARE FEET                  26049
GROSS SQUARE FEET                 26049
YEAR BUILT                            0
TAX CLASS AT TIME OF SALE             0
BUILDING CLASS AT TIME OF SALE        0
SALE PRICE                        14176
SALE DATE                             0
dtype: int64

In [ ]:
#kita drop data yang null
df.dropna(inplace=True)

In [ ]:
df.isnull().sum()

BOROUGH                           0
NEIGHBORHOOD                      0
BUILDING CLASS CATEGORY           0
TAX CLASS AT PRESENT              0
BLOCK                             0
LOT                               0
BUILDING CLASS AT PRESENT         0
ADDRESS                           0
APARTMENT NUMBER                  0
ZIP CODE                          0
RESIDENTIAL UNITS                 0
COMMERCIAL UNITS                  0
TOTAL UNITS                       0
LAND SQUARE FEET                  0
GROSS SQUARE FEET                 0
YEAR BUILT                        0
TAX CLASS AT TIME OF SALE         0
BUILDING CLASS AT TIME OF SALE    0
SALE PRICE                        0
SALE DATE                         0
dtype: int64

## Exploratory data analysis

Now, let's get a simple descriptive statistics with `.describe()` function for `COMMERCIAL UNITS` features.

In [ ]:
df[df['COMMERCIAL UNITS']==0].describe()

BOROUGH         BLOCK  ...    YEAR BUILT    SALE PRICE
count  44504.000000  44504.000000  ...  44504.000000  4.450400e+04
mean       3.299254   5103.551681  ...   1803.763302  7.563546e+05
std        0.897414   3466.307329  ...    507.825487  2.950508e+06
min        1.000000      1.000000  ...      0.000000  0.000000e+00
25%        3.000000   2299.000000  ...   1920.000000  1.267500e+05
50%        3.000000   4753.500000  ...   1930.000000  4.790000e+05
75%        4.000000   7135.000000  ...   1962.000000  7.990000e+05
max        5.000000  16319.000000  ...   2017.000000  3.450000e+08

[8 rows x 11 columns]

Let us try to understand the columns. Above table shows descriptive statistics for the numeric columns.

- There are zipcodes with 0 value
- Can block/lot numbers go up to 16322?
- Most of the properties have 2 unit and maximum of 1844 units? The latter might mean some company purchased a building. This should be treated as an outlier.
- Other columns also have outliers which needs further investigation.
- Year column has a year with 0
- Most sales prices less than 10000 can be treated as gift or transfer fees.

Now, let's get a simple descriptive statistics with `.describe()` function for `RESIDENTIAL UNITS` features.

Function below are graded function. (1 Points)

In [ ]:
# Write your function below
df[df['RESIDENTIAL UNITS']==0].describe()

# Graded-Funtion Begin (~1 Lines)

# Graded-Funtion End

BOROUGH         BLOCK  ...   YEAR BUILT    SALE PRICE
count  8308.000000   8308.000000  ...  8308.000000  8.308000e+03
mean      2.853033   4465.301517  ...  1514.548146  2.509375e+06
std       0.713720   2952.551631  ...   814.794695  3.078971e+07
min       1.000000      1.000000  ...     0.000000  0.000000e+00
25%       3.000000   1944.000000  ...  1907.750000  0.000000e+00
50%       3.000000   4586.000000  ...  1937.000000  2.390000e+05
75%       3.000000   6570.750000  ...  1961.000000  6.500000e+05
max       5.000000  16315.000000  ...  2017.000000  2.210000e+09

[8 rows x 11 columns]

Write your findings below:



*   ada zipcode yang memiliki 0 value
*   Untuk Residential memiliki rata-rata SALE PRICE unit lebih dari 1000000



Use `.value_counts` function to count total value of `BOROUGH` features. Expected value:

4    26548\
3    23843\
1    18102\
5     8296\
2     6994\
Name: BOROUGH, dtype: int64

In [ ]:
# Write your syntax below

df_borough = df['BOROUGH'].value_counts()
df_borough

3    23843
4    11292
2     6994
5     5311
1     1013
Name: BOROUGH, dtype: int64

From here, we can calculate the mean for each Borough. Use `.mean()` function to calculate mean.

Function below are graded function. (1 Points)

In [ ]:
# Write your function below
df_borough.mean()
# Graded-Funtion Begin (~1 Lines)



# Graded-Funtion End



9690.6

From here, we can calculate the median for each Borough. Use `.median()` function to calculate median.

Function below are graded function. (1 Points)

In [ ]:
# Write your function below
df_borough.median()
# Graded-Funtion Begin (~1 Lines)

# Graded-Funtion End

6994.0

From here, we can calculate the mode for each Borough.

Function below are graded function. (1 Points)

In [ ]:
# Write your function below
df_borough.mode()
# Graded-Funtion Begin (~1 Lines)

# Graded-Funtion End

0     1013
1     5311
2     6994
3    11292
4    23843
dtype: int64

From here, we can calculate the Range for each Borough.

Function below are graded function. (1 Points)

In [ ]:
# Write your function below
np.ptp(df_borough)
# Graded-Funtion Begin (~1 Lines)

# Graded-Funtion End

22830

From here, we can calculate the Variance for each Borough.

Function below are graded function. (1 Points)

In [ ]:
# Write your function below
df_borough.var()
# Graded-Funtion Begin (~1 Lines)

# Graded-Funtion End

76152049.3

From here, we can calculate the SD for each Borough.

Function below are graded function. (1 Points)

In [ ]:
# Write your function below
df_borough.std(ddof=1)
# Graded-Funtion Begin (~1 Lines)

# Graded-Funtion End

8726.51415514809

Contoh untuk CI
# Berapa % property di NYC yang diatas 3000000

In [ ]:
# Cari jumlah nya dulu
df_priceover = df[df['SALE PRICE'] > 3000000]
print(f"jumlah data properti: {df.shape}")
print(f"jumlah properti diatas 3000000  : {df_priceover.shape}")


jumlah data properti: (48453, 20)
jumlah properti diatas 3000000  : (1948, 20)


# Now we can analyze **Confidence Intervals** below.

Function below are graded function. (1 Points)

In [ ]:
# Write your function below

# Graded-Funtion Begin
p = 1948/48453
jumlah_data = 83783
z = 1.96
standard_error = np.sqrt(p*(1-p)/jumlah_data)
error = z*standard_error

np.array((p - error, p+error))*100




# Graded-Funtion End

array([3.88737555, 4.15340624])

# Maka dari itu diketahui bahwa Confidence Intervals jumlah properti diatas SALE PRICE 3000000 adalah 3 hingga 4 persen




# Make your **Hypothesis Testing** below

Function below are graded function. (1 Points)

# Research Question nya apakah NEIGBORHOOD ALPHABET CITY memiliki SALE PRICE rata-rata yang lebih tinggi secara signifikan daripada WOODROW?


In [ ]:
df_Neigh = df[["NEIGHBORHOOD","SALE PRICE"]]
df_alpha= df_Neigh[df_Neigh['NEIGHBORHOOD'] == 'ALPHABET CITY']
df_alpha

NEIGHBORHOOD  SALE PRICE
0    ALPHABET CITY   6625000.0
3    ALPHABET CITY   3936272.0
4    ALPHABET CITY   8000000.0
6    ALPHABET CITY   3192840.0
9    ALPHABET CITY  16232000.0
11   ALPHABET CITY  10350000.0
22   ALPHABET CITY  11900000.0
73   ALPHABET CITY  28000000.0
75   ALPHABET CITY  52625000.0
172  ALPHABET CITY   3300000.0
173  ALPHABET CITY   7215000.0
174  ALPHABET CITY   4750000.0
195  ALPHABET CITY   3700000.0

In [ ]:
df_wood = df_Neigh[df_Neigh['NEIGHBORHOOD'] == 'WOODROW']
df_wood


NEIGHBORHOOD  SALE PRICE
84417      WOODROW    640000.0
84418      WOODROW    664000.0
84422      WOODROW    470000.0
84423      WOODROW    470000.0
84424      WOODROW    465000.0
...            ...         ...
84543      WOODROW    450000.0
84544      WOODROW    550000.0
84545      WOODROW    460000.0
84546      WOODROW  11693337.0
84547      WOODROW     69300.0

[93 rows x 2 columns]

In [ ]:
n1 = len(df_alpha)
mu1 = df_alpha["SALE PRICE"].mean()
sd1 = df_alpha["SALE PRICE"].std()

(n1, mu1, sd1)

(13, 12294316.307692308, 13944610.632686878)

In [ ]:
n2 = len(df_wood)
mu2 = df_wood["SALE PRICE"].mean()
sd2 = df_wood["SALE PRICE"].std()

(n2, mu2, sd2)

(93, 668532.5161290322, 1173676.5327400514)

In [ ]:
# Write your function below



# Graded-Funtion Begin

sm.stats.ztest(df_alpha["SALE PRICE"].dropna(), df_wood["SALE PRICE"].dropna(),alternative='two-sided')



# Graded-Funtion End

(8.072678787185998, 6.877236123189402e-16)

Write your final conclusion below.

Your conclusion below are graded. (1 Points)

# CONCLUSION


*   Dari Hasil Confindence Intervals jumlah properti diatas SALE PRICE 3000000 adalah 3 hingga 4 persen
*  Dari Hasil Hypothesis Testing harga properti NEIGHBORHOOD Alphabet City memiliki rata-rata lebih tinggi dibandingkan Woodrow

